feat(DEV-4699): Provide an ability to edit an invoice

MadMed677 opened the pull request:
![ar-edit](https://github.com/team-monite/monite-sdk/assets/6944144/d65734ec-ee58-421f-8673-cf6623ab153d)

Works on `dev` stand

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/EntitySection.tsx:
```diff
@@ -35,7 +35,24 @@ const detailsGridItemProps = {
+  'purchase_order',
+] as const;
+
+interface IEntitySectionProps extends ISectionGeneralProps {
```
Maybe we should follow the MUI style of naming interfaces?
`IEntitySectionProps` => `EntitySectionProps`

MadMed677 replied:
Hm, why? :)
We always use these options. How is our code style related to MUI code style?

radist2s replied:
I don't recall us adopting any convention, accordingly, we follow the MUI style.

MadMed677 replied:
I don't see any convention on our side. I personally use this format for all interfaces

nekedos replied:
@MadMed677 Can I personally ask you to avoid such naming and follow the MUI style?

radist2s replied:
I would prefer to keep MUI style in all naming.

MadMed677 replied:
Updated

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/useExistingInvoiceDetails.tsx:
```diff
@@ -133,16 +161,26 @@ export function useExistingInvoiceDetails({
   }, [deliveryMethod, receivable?.status]);
 
+  const isEditButtonVisible = useMemo(() => {
+    return (
```
Probably, we have to check the user's permissions too.

MadMed677 replied:
We don't handle permissions for receivables (at all) right now, unfortunately :(

radist2s replied:
Do you mean there is no support on the backend for differentiating permissions?

MadMed677 replied:
On BE side it's supported I guess, but on FE side there is no implementation of it

radist2s replied:
Not, sure, but probably [this is it](https://github.com/team-monite/monite-sdk/blob/bde068d520a604abf069d910e48711a36fe58039/packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/ExistingReceivableDetails.tsx#L77-L81).

MadMed677 replied:
`ExistingReceivableDetails` works only for `Quotes` & `Credit Notes` it has outdated UI and have been made more than a year ago and shouldn't be supported
https://github.com/team-monite/monite-sdk/blob/1cba52101f1691de8f7027b27da07e260b6b9147/packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/ExistingReceivableDetails.tsx#L60

All improvements are made only for `Invoices`. Currently `Invoices` doesn't support backend permissions

I don't want to add to this task which is called `Provide an ability to edit invoice` functionality to check permissions because it's out of scope of this task.

I created a task to support BE permissions for a new component ([task link](https://monite.atlassian.net/browse/DEV-10818)). Will it work for you in that case? To make it in a separate task

radist2s replied:
Sure, btw, which actions are not supported?

MadMed677 replied:
What do you mean? Permissions? All. All permissions are not supported on FE side for a new component

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/EditInvoiceDetails.tsx:
```diff
@@ -0,0 +1,287 @@
+      <DialogContent>
+        <FormProvider {...methods}>
+          <form
+            id="receivablesDetailsForm"
```
It would be better to replace `receivablesDetailsForm` with the unique ID:
```ts
const formId = `Monite-Form-${useId()}`
```

Probably, you can add similar changes to [EditInvoiceDetails](https://github.com/team-monite/monite-sdk/blob/9ceb783e8a09c1b63355e25675b34fade03f16fe/packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/EditInvoiceDetails.tsx#L182)

MadMed677 replied:
Hm, but why? We always use specific form ids

radist2s replied:
There may be a situation the `id` on the customer's page will conflict with the non-unique `id` of `receivablesDetailsForm`

MadMed677 replied:
Let's add this topic to FE sync. Because it's a general change and if it really possible to have any problem with it we could change it but globally for all forms then

radist2s replied:
It seems better to avoid obvious mistakes in advance than to correct them in the future.

nekedos replied:
@MadMed677, By the way, why do we need to use `id`?

MadMed677 replied:
@nekedos When we submit the form we have to provide which form we want to submit. 
Example: https://github.com/team-monite/monite-sdk/blob/61ee304aa3066640b0a49c78d735cbdf0ac4d1b5/packages/sdk-react/src/components/approvalPolicies/ApprovalPolicyDetails/ApprovalPolicyDetailsForm/ApprovalPolicyDetailsForm.tsx#L274

MadMed677 replied:
@radist2s updated

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/ExistingInvoiceDetails.tsx:
```diff
@@ -221,6 +235,14 @@ export const ExistingInvoiceDetails = (
                   >{t(i18n)`Delete`}</Button>
                 )}
+                {buttons.isEditButtonVisible && (
+                  <Button
```
Alternatively, you can add permissions checking right here. Btw, _Delete_ action is missing permission checking too. Check out [an example](https://github.com/team-monite/monite-sdk/commit/bde068d520a604abf069d910e48711a36fe58039#diff-4b85029f45ba03ebdfa44b46e4f56a24dfa7c864daac70d2aaeafc61b6f1faeaR72-R76)!

MadMed677 replied:
We don't support permissions right now

radist2s commented:
The same cool dialog will not be shown, if you press <kbd>Esc</kbd> 😞

<img width="1426" alt="image" src="https://github.com/team-monite/monite-sdk/assets/725645/52e2dd87-30be-45c2-9a5f-806a14d8bb99">

radist2s commented:
I think I'm confused.

- In the first screenshot we see VAT ID `27088af0`

  <img width="1421" alt="image" src="https://github.com/team-monite/monite-sdk/assets/725645/48437d1a-79d5-45e0-8edc-a067e6ebd297">

- In the next VAT ID is empty but `27088af0` refers to Tax ID.

  <img width="1428" alt="image" src="https://github.com/team-monite/monite-sdk/assets/725645/13bc03d7-c0b2-4f5d-b7a7-00e1f25e7a11">

radist2s commented:
If update an Invoice, you will be returned to the Invoice Details dialog, but PDF will not load immediately, as it is not yet ready on the server. Is possible to specify somehow better information about PDF status? Also, the alignment of the message seems broken.

<img width="1419" alt="image" src="https://github.com/team-monite/monite-sdk/assets/725645/7ad7af0c-ddd2-4c80-92f3-8b82fb1dabc2">

radist2s commented:
@MadMed677, huge work! Thank you!

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/EditInvoiceDetails.tsx:
```diff
@@ -0,0 +1,287 @@
+                },
+              };
+
+              updateReceivableLineItems.mutate(lineItems, {
```
If possible do not update `lineItems` if not changed, that would be cool.

MadMed677 replied:
Yep, it sounds good. But do you know how to check it? Because as I know `react-hook-form` will create a new array of `line_items` and if we check `current.lineItems === server.lineItems` we will have always `false` because it will be a new array.

So the only one way is to check it by deep check (check all fields by going through all items) which I think will be worse than current one.

Maybe if we could handle `isDirty` but not for the whole form but only for `line_items` we could avoid any checks but I didn't find any options to do so

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/EditInvoiceDetails.tsx:
```diff
@@ -0,0 +1,287 @@
+                      format(values.fulfillment_date, 'yyyy-MM-dd')
+                    : undefined,
+                  /** !!! Note !!! Backend is not supported to edit `purchase_order` so we have to remove it */
+                  // purchase_order: values.purchase_order || undefined,
```
Is the backend going to update the OpenAPI Document according to this issue?

MadMed677 replied:
They already have it :)

I just realized that during implementation. They have to fix it and support these fields but we never know when

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/sections/EntitySection.tsx:
```diff
@@ -50,13 +67,31 @@ export const EntitySection = ({ disabled }: ISectionGeneralProps) => {
+    return allFields.filter((field) => field !== 'purchase_order');
+  }, [hidden]);
+
+  const gridItemProps = useMemo(() => {
```
💋

MadMed677 commented:
> the Dialog will also be displayed (as if the form becomes dirty with no real changes).

Yep, but it's impossible to fix it :( That's how react-hook-form works. It sets `isDirty: true` when you change anything and even if it's the same value react-hook-form doesn't handle it

MadMed677 commented:
> The same cool dialog will not be shown, if you press Esc 😞
> 
> <img alt="image" width="1426" src="https://private-user-images.githubusercontent.com/725645/327827076-52e2dd87-30be-45c2-9a5f-806a14d8bb99.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzMzQxMDMsIm5iZiI6MTcxNTMzMzgwMywicGF0aCI6Ii83MjU2NDUvMzI3ODI3MDc2LTUyZTJkZDg3LTMwYmUtNDVjMi05YTVmLTgwNmExNGQ4YmI5OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxMFQwOTM2NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZmE1ZmQyYmIxMDlkYTNhZTZlODBlMDA4YTNhMDFlZDM0MDE1NDgxODVjZjIzNmI4ZWM2MTM5MTJhNjhlNjhkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.tt6hQ04uvTghh4wdKGOlnFBuLRlA0di6XWeT7J4MGVo">
> **Btw**, even if I didn't change anything, ~but just expanded the Select and closed it,~ the Dialog will also be displayed (as if the form becomes dirty with no real changes).

I don't know how it should work. Undortunately design team doesn't support it. Right now `ECS` works only when you want to close pop-up but current page is not a pop-up window just another _route_ to edit an invoice. If you check any other entity (payables, counterparts, etc) they will work the same

MadMed677 commented:
> I think I'm confused.
> 
> * In the first screenshot we see VAT ID `27088af0`
>   <img alt="image" width="1421" src="https://private-user-images.githubusercontent.com/725645/327828876-48437d1a-79d5-45e0-8edc-a067e6ebd297.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU1OTE5MzAsIm5iZiI6MTcxNTU5MTYzMCwicGF0aCI6Ii83MjU2NDUvMzI3ODI4ODc2LTQ4NDM3ZDFhLTc5ZDUtNDVlMC04ZWRjLWEwNjdlNmViZDI5Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxM1QwOTEzNTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMzM5ZjBkY2IwMTNiNDUyOWFlNjY2Mzg1MjM3NjhkNTY2MDM2MTI5ZmYxMzljNTI4MzA3M2QxZTM3NTY1OGYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.qCCzmc462NetDuzERMwKXepIgx0TB-X1cgdqIpX-CsM">
> * In the next VAT ID is empty but `27088af0` refers to Tax ID.
>   <img alt="image" width="1428" src="https://private-user-images.githubusercontent.com/725645/327829103-13bc03d7-c0b2-4f5d-b7a7-00e1f25e7a11.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU1OTE5MzAsIm5iZiI6MTcxNTU5MTYzMCwicGF0aCI6Ii83MjU2NDUvMzI3ODI5MTAzLTEzYmMwM2Q3LWMwYjItNGY1ZC1iN2E3LTAwZTFmMjVlN2ExMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxM1QwOTEzNTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iMTJlYmY0NTBmMTFkYmQ1N2FkZmJjZTM2YzFkZTMyNDJkYWI4NzVjZmZmYzVhNWZkZTI4NmU5MWRlYTJmY2ZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Wqe0m63Cq_CTNpFyAQgdUrnCEMdsPBuHpbMZpcaKW4I">
> * Finally, updated VAT ID
>   <img alt="image" width="571" src="https://private-user-images.githubusercontent.com/725645/327830642-bc3afb5c-b6b9-4af7-a9f9-f3d5940fde49.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU1OTE5MzAsIm5iZiI6MTcxNTU5MTYzMCwicGF0aCI6Ii83MjU2NDUvMzI3ODMwNjQyLWJjM2FmYjVjLWI2YjktNGFmNy1hOWY5LWYzZDU5NDBmZGU0OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxM1QwOTEzNTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNTYwMzZjNTQ3NThhNjJjODViM2M2ZmVjYmUxMDJiZTE1YTlhMGZlNTMzZTJlMDI0Y2YwNzVjMjY3NmFiN2ExJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.3j7u_YPzTMP1q1U6PIqPldpfhFMA_QdNE1FEnjS9FFI">

Updated

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/CreateReceivable/CreateReceivables.tsx:
```diff
@@ -108,6 +108,9 @@ const CreateReceivablesBase = (props: InvoiceDetailsCreateProps) => {
   }
 
+  // eslint-disable-next-line lingui/no-unlocalized-strings
+  const formName = 'Monite-Form-receivablesDetailsForm';
```
Could you please refactor it with the `Monite-Form-receivablesDetailsForm-useId()`?

radist2s replied:
You're quite right, the probability of having 2 `Monite-Form-receivablesDetailsForm` identifiers on the page is low. But why think about probabilities when it's better to use [the standard approach](https://react.dev/reference/react/useId#why-is-useid-better-than-an-incrementing-counter) that React offers

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/EditInvoiceDetails.tsx:
```diff
@@ -0,0 +1,289 @@
+  } = methods;
+
+  useEffect(() => {
+    setValue('default_shipping_address_id', counterpartShippingAddress?.id);
```
I suppose, this is the reason why the form state is always `touched`. It would be better to avoid using such hacks.

MadMed677 replied:
Updated fc7432639ab94a4a829676230bb7df6fe0c94fd5

radist2s commented on packages/sdk-react/src/components/receivables/InvoiceDetails/ExistingInvoiceDetails/components/EditInvoiceDetails.tsx:
```diff
@@ -0,0 +1,289 @@
+    updateReceivableLineItems.isPending || updateReceivable.isPending;
+
+  // eslint-disable-next-line lingui/no-unlocalized-strings
+  const formName = 'Monite-Form-receivablesDetailsForm';
```
Would be better to refactor as

```
const formName = `Monite-Form-receivablesDetailsForm-${useId()}`
```

MadMed677 replied:
Updated https://github.com/team-monite/monite-sdk/commit/eab3d83aca9da2d41e5cc17430d92f03da2378f0
